[Nullability Annotations to Java Classes] Replace `findViewById` with `ViewBinding` for Comment (`relatively safe` -> `risky`)

ParaskP7 opened the pull request:
Parent #18906
Relates To #19217

This PR's main focus is to replace `findViewById ` with `ViewBinding` for the [CommentDetailFragment.java](https://github.com/wordpress-mobile/WordPress-Android/blob/8ce321d5e9f6eaed0adf8b2128da720d5e365292/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java#L4) class.

Additionally, this PR is also dealing with adding any missing nullability annotations to this activity class, focusing mainly on its fields, and, any effected neighbour classes. But first, all existing warnings on this class are being resolved/suppressed so as to make it easier to deal with any missing nullability checks (with help from the IDE), especially when a `@Nullable` annotation is added.

FYI: Although this change is mainly `relatively safe`, it is also quite `risky`. 
- The `relatively safe` aspect of this PR means that although there are compile-time changes associated with this change, and it needs testing, the changes included in this PR are very targeted and specific to one screen, and one screen only. Thus, if this screen is tested appropriately, it should be safe to merge this to `trunk`.
- The `risky` aspect of this PR means that although the changes included in this PR are done progressively and in a way that is safe, the complexity of this `fragment` class, along with the fact that they were more than a dozen of NPE that got addressed, this could easily translate to some behaviors changing along they way, which might have somehow affected how the screen behaves on certain flows. As such, this change needs thorough testing prior merging to `trunk`.

PS: I really suggest doing a commit-by-commit code review and try to follow/assert that each commit is correctly applied.

-----

PS: @irfano I added you as the main reviewer, randomly so, since I just wanted someone from the Jetpack/WordPress mobile team to be aware of and sign-off on that change for JP/WPAndroid. Feel free to merge this PR directly yourself if you deem so.

-----

`findViewById ` -> `ViewBinding`

1. [Replace findviewbyid with viewbinding for comment detail (1/2)](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/672e09b24477ca4f7d7998caa5b481e30606c64e)
2. [Replace findviewbyid with viewbinding for comment detail (2/2)](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/bc0fb45cacec39a307c86237a38b91251031904c)
3. [Use non-null viewbinding method param for comment detail (1/3)](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/950250fdc5bc7979f2dca28c37072e0d6dea8eb8)
4. [Use non-null viewbinding method param for comment detail (2/3)](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/f36fea8d79d8f96482dd587b7817b8746caa3b93)
5. [Use non-null viewbinding method param for comment detail (3/3)](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/b97bd36d1d195a49ee5b6fbcee5eb709ec876ed8)

Nullability Annotation List:

1. [Add missing nullability annotations to sdk override methods](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/71a22bc8f2671ecf58a6bd3bf4c82acfee9b0194)
2. [Add missing nl-a to get note method on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/c026d3ac7e2dc409058d12f823ca88ea820d46b8)
3. [Add missing nl-a to comment field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/ca577fea663c75470b71837c0f549977c6a16bc3)
4. [Use nn-a comment parameter for show/set post title](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/cf8842663ac4f0ec39352601620202f04d9f26d6)
5. [Change to nn-a binding parameter on set post title](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/7d1bba7f1f29d9757def5f407a4bd30cb88f0b81)
6. [Use nl-a comment parameter for set reply unique id](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/cf8f6571c0803f29fe7cee5a8693d085e48a0511)
7. [Utilize nl-a on node field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/0d43200e11ff531fff9577f8a02b8512bc641acf)
8. [Add missing nl-a to site field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/9e0845b417583b398a15ab072fe1574ab03c64a1)
9. [Add missing nl-a to suggestion adapter field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/dcad265ffb83becf1b03915e64845379c9edc191)
10. [Add missing nl-a to suggestion service connection manager fld](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/6d88036416b7929e00435790ac9ff6af7f4bf398)
11. [Add missing nl-a to restored reply text field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/2cb161926c74e6f27b2b99f65c4cdfd07eaea1ab)
12. [Add missing nl-a to restored note id field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/cc449791e589bffa1e9732fb18d9e969150e8d92)
13. [Add missing nl-a to previous status field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/647ba6c58984bb0cde3d4ce04f10e12a1fe090ab)
14. [Add missing nl-a to notifications detail list fragment field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/969b02dca562fe3086888899a65b9da621243d74)
15. [Add missing nl-a to on post click listener field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/be7cb0c550baf589ca80c8217566005397d5043a)
16. [Add missing nl-a to on comment action listener field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/a941ea405006b40e42bdc7ba9802e7e0b4abe128)
17. [Add missing nl-a to on note comment action listener field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/8c5d1de1a3168009c84837accd04efbe3701f823)
18. [Add missing nl-a to comment source field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/d6e7f9e49ffce58ae424d8e2a527d5c817929b28)
19. [Add missing nn-a to enabled actions field](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/28d993a422bd2c7b005f821582a8602141324063)

Warnings Resolution List:

1. [Create missing switch branch for track moderation event](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/880c02ebc778ac502a058c73e4aa38dc25a87575)
2. [Create missing switch branches for update status views](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/641211bb15f45dbaf4d180ba9b68255d33239f47)
3. [Create missing switch branch for announce comment status](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/59e42c2cee4404d9758d96ff0dbb882f3ddd752d)
4. [Replace get activity with require activity on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/da9ff13f7075bf633c18c5a8d959e0d68db73d69)
5. [Replace get arguments with require arguments on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/c3a49172de4cd0e72c45252dc053dfd93e0b64dd)
6. [Make normal opacity a static final field on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/8bbc30df2201bebecc67045483038d6b0e7da329)
7. [Remove unnecessary suggestion adapter null check condition](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/9a7fdba471a41d13e0384b8645539984d37b2dab)
8. [Guard nullable site model variable on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/624dd8acaf4ce74385b8e20abc4cf5c7a1025513)
9. [Guard nullable note field on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/c9ae16b8548ac2fb5149563697c0a4e3e65a94f5)
10. [Guard nullable comment identifier variable on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/a2e0d09d236df12647cc72982e994c639a5fc3b1)
11. [Access static method via class color utils reference](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/414b6b45c037c054c1577aada380880bee4fc045)
12. [Guard nullable comment field on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/c0fbc0475df0e566994375aa7290ae4a7eeb641c)
13. [Remove unnecessary return on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/4e4b039672f8e2074f2590094323e7e301380f9e)
14. [Simplify if/else for show more menu on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/74726117fdcb5a6494020dd9084939258a53f357)

Warnings Suppression List:

1. [Suppress deprecation warnings on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/1dbc843c611f6455bba43f92ddc8f7a08aa9b9a7)
2. [Suppress same parameter value on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/0dfac3b0bf90fd3c625f5464c282574f56c9e842)

Refactor List:

1. [Remove unnecessary is added check for show comment](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/3b0e34155ce8d0831ac7b44c426efe8ba084ad17)
2. [Move on destroy at the bottom close to on destroy view method](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/c3b5a83c3a7d7c753c744fa00f08346e4b2b597a)
3. [Split show comment into two for nullable and non-null comment](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/40da17b76dfc6c32b0fd87ae8c68b69671a46e27)
4. [Reorder site param on top of comment for set comment method](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/d3d796334ad6a3a7882c63bc074c27e10548d04e)

Documentations List:

1. [Fixed initialized type on comment detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/fa969d5d2937dbea515d0582618c7cdb16c6e2c8)
2. [Reformat add detail fragment comment](https://github.com/wordpress-mobile/WordPress-Android/pull/19232/commits/c08c44e5f119480abb33e94c61ce2a008c65c59e)

-----

## To test:

1. Thoroughly smoke test any comment related functionality on both, the WordPress and Jetpack apps, and see if everything is working as expected.
2. In addition to the above smoke test, you can expand the below and follow the inner and more explicit test steps within:

<details>
    <summary>1. Comment Details Screen from Comments Screen [CommentDetailFragment.java]</summary>

ℹ️ This test applies to both, the `Jetpack` and `WordPress` app.

- Go to `Comments` screen and tap on any comment.
- Verify that the `Comment Details` screen is shown and functioning as expected.
- For example, try replying, approving, marking as spam, liking, trashing, editing or copying a comment's link address. Additionally, make sure to verify that you can still swipe left-right to navigate between comments.

</details>

<details>
    <summary>2. Comment Details Screen from Notification Tab [CommentDetailFragment.java]</summary>

ℹ️ This test applies to the `Jetpack` app only.

- Go to `Notifications` tab, then its `COMMENTS` sub-tab, and tap on any comment.
- Verify that the `Comment Details` screen is shown and functioning as expected.
- For example, try replying to a comment. Additionally, make sure to verify that you can still swipe left-right to navigate between comments.

</details>

FYI: While I was testing the `Comment` screen I noticed a couple of bugs. These bugs are not caused by any of the changes included in this PR, and, can be easily reproduced on `trunk` as well. For more info see: #19236 & #19237

PS: I also noticed another bug (on `trunk` as well), but was only able to reproduce once (and then couldn't). I am going to write that down here just in case you notice that bug as well. Here goes: `When editing a comment and you end up saving a comment that was already edited with the exact same set of characters in the previous edit, then this new edit is not saved, instead it is being reset to the previous edit.` 🤷 🤔 🤷

-----

## Regression Notes

1. Potential unintended areas of impact

    - Potential screen and flow misbehaviour, especially due to the fact that this `fragment` applies to both, the `Comments` and `Notifications` screens, via the very convoluted `Comment` and `Node` model flows correspondingly, which are used interchangingly during each screen's initialization process. As such, this change needs thorough testing to verify correctness.

18. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

34. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

ParaskP7 committed:
Refactor: Replace findviewbyid with viewbinding for comment detail (1/2)

This change includes the inner 'mReplyBinding' binding, which is an
included layout within the 'comment_detail_fragment.xml' file, with the
'layout_comment_box' id, which points to the
'reader_include_comment_box.xml' layout file.

FYI: In order to reference the included 'layout_comment_box' id, the
'layout_container' id needed to be created on the top level view of the
the 'reader_include_comment_box.xml' layout file.

ParaskP7 committed:
Refactor: Replace findviewbyid with viewbinding for comment detail (2/2)

This change includes the outer 'mActionBinding' binding, which is added
to the main binding's 'commentContentContainer' layout programmatically,
which points to the 'comment_action_footer.xml' layout file.

FYI: The 'layout_buttons' id is already created on the top level view
of the the 'comment_action_footer.xml' layout file. As such, referencing
this is already possible.

ParaskP7 committed:
Refactor: Use non-null viewbinding method param for comment detail (1/3)

This is done in order to avoid using the 'if (mBinding != null) { ... }'
check everywhere and only use that where absolutely necessary, which
usually is only needed when a 'binding' is required via an SDK related
'@Override' method, that is, it being the 'binding' starting point.

FYI: This way, when this class gets converted to Kotlin (in the future),
instead of using the 'binding' as a parameter to a method, the same
method will get converted to a 'binding' extension function, and then,
everything will become much more straightforward. For what is worth,
this pattern is already used as such with Kotlin classes. For example,
see 'PreviewImageFragment.kt' class and its 'initializeViews()'
extension function.

PS: An exception to this were the 'setComment(...)' methods, and the
fact that the non-null 'binding' viewbinding wasn't added as a parameter
to those methods. This was done because these methods are also called
from within 'onCreate(...)' SDK related '@Override' fragment lifecycle
method, which is called before the 'onCreateView(...)' SDK related
'@Override' fragment lifecycle method. As such, and because the
'mBinding' field is only set during the 'onCreateView(...)' SDK related
'@Override' fragment lifecycle method, it would always be null during
the 'onCreate(...)' SDK related '@Override' fragment lifecycle method,
causing a logical initialization change on this comment detail screen,
which will most probably cause an issue due to the fact that
the 'mComment', 'mSite' and 'mIsUsersBlog' fields would not be set,
something that is actually done via the 'setComment(...)' method itself
and before it needing to check and access a non-null 'mBinding'. Same
logical thinking and process was applied to 'setNote(...)' and the fact
that the 'mNote' and 'mSite' fields would not be set otherwise.

ParaskP7 committed:
Refactor: Remove unnecessary is added check for show comment

The 'isAdded()' check is already available within the 'showComment(...)'
method itself, on the very first statement of the method, which returns
immediately if 'isAdded()' is not met. As such, having any client that
is calling 'showComment(...)' check that again is unnecessary, plus,
this unnecessarily increases the complexity of the class overall.

FYI: Actually, the 'onStart(...)' SDK related '@Override' fragment
lifecycle method is already depending on this fact.

ParaskP7 committed:
Refactor: Use non-null viewbinding method param for comment detail (2/3)

This is done in order to avoid using the
'if (mReplyBinding != null) { ... }' check everywhere and only use that
where absolutely necessary, which usually is only needed when a
'replyBinding' is required via an SDK related '@Override' method, that
is, it being the 'replyBinding' starting point.

FYI: This way, when this class gets converted to Kotlin (in the future),
instead of using the 'replyBinding' as a parameter to a method, the
same method will get converted to a 'replyBinding' extension function,
and then, everything will become much more straightforward. For what is
worth, this pattern is already used as such with Kotlin classes. For
example, see 'PreviewImageFragment.kt' class and its 'initializeViews()'
extension function.

PS: As per the previous commit message description, an expeption to this
were the 'setComment(...)' and 'setNote(...)' methods, for the very same
reasons.

ParaskP7 committed:
Refactor: Use non-null viewbinding method param for comment detail (3/3)

This is done in order to avoid using the
'if (mActionBinding != null) { ... }' check everywhere and only use that
where absolutely necessary, which usually is only needed when a
'actionBinding' is required via an SDK related '@Override' method, that
is, it being the 'actionBinding' starting point.

FYI: This way, when this class gets converted to Kotlin (in the future),
instead of using the 'actionBinding' as a parameter to a method, the
same method will get converted to a 'actionBinding' extension function,
and then, everything will become much more straightforward. For what is
worth, this pattern is already used as such with Kotlin classes. For
example, see 'PreviewImageFragment.kt' class and its 'initializeViews()'
extension function.

PS: As per the previous commit message description, an expeption to this
were the 'setComment(...)' and 'setNote(...)' methods, for the very same
reasons.

ParaskP7 committed:
Refactor: Move on destroy at the bottom close to on destroy view method

ParaskP7 committed:
Analysis: Add missing nullability annotations to sdk override methods

ParaskP7 committed:
Analysis: Create missing switch branch for track moderation event

Warning: "'switch' statement on enum type
'org.wordpress.android.fluxc.model.CommentStatus' misses case
'UNREPLIED'"

ParaskP7 committed:
Analysis: Create missing switch branches for update status views

Warning: "'switch' statement on enum type
'org.wordpress.android.fluxc.model.CommentStatus' misses cases:
'DELETED', 'ALL', 'UNREPLIED', 'UNSPAM', 'UNTRASH'"

ParaskP7 committed:
Analysis: Create missing switch branch for announce comment status

Warning: "'switch' statement on enum type
'org.wordpress.android.fluxc.model.CommentStatus' misses case
'UNREPLIED'"

ParaskP7 committed:
Analysis: Suppress deprecation warnings on comment detail

Warnings (Feature Related):
- "Deprecated member 'CommentDetailFragment' is still used"
- "'org.wordpress.android.ui.comments.unified.CommentsStoreAdapter'
is deprecated"
- "'org.wordpress.android.ui.comments.CommentDetailFragment'
is deprecated"
- "'org.wordpress.android.ui.comments.EditCommentActivity'
is deprecated"

Warnings: (SDK Related):
- "'setHasOptionsMenu(boolean)' is deprecated"
- "'startActivityForResult(android.content.Intent, int)' is deprecated"
- "'hideSoftInput(android.widget.EditText)' is deprecated"
- "'showSoftInput(android.widget.EditText)' is deprecated"

ParaskP7 committed:
Analysis: Replace get activity with require activity on comment detail

Warnings:
- "Method invocation 'getApplication' may produce
'NullPointerException'"
- "Argument 'getActivity()' might be null"
- "Method invocation 'invalidateOptionsMenu' may produce
'NullPointerException'"
- "Method invocation 'getSystemService' may produce
'NullPointerException'"

ParaskP7 committed:
Analysis: Replace get arguments with require arguments on comment detail

Warnings:
- "Method invocation 'getSerializable' may produce
'NullPointerException'"
- "Method invocation 'getLong' may produce
'NullPointerException'"
- "Method invocation 'getInt' may produce
'NullPointerException'"
- "Method invocation 'getString' may produce
'NullPointerException'"

ParaskP7 committed:
Analysis: Make normal opacity a static final field on comment detail

Warning: "Field 'mNormalOpacity' may be 'final'"

ParaskP7 committed:
Analysis: Remove unnecessary suggestion adapter null check condition

Warning: "Condition 'mSuggestionAdapter != null' is always 'true'"

ParaskP7 committed:
Analysis: Guard nullable site model variable on comment detail

Warning: "Argument 'site' might be null"

ParaskP7 committed:
Analysis: Add missing nl-a to get note method on comment detail

FYI: 'nl-a' stands for 'nullable annotation'.

Warning: "Getter for @Nullable field might be annotated @Nullable
itself"

ParaskP7 committed:
Analysis: Guard nullable note field on comment detail

Warnings:
- "Method invocation 'getId' may produce
'NullPointerException'"
- "Method invocation 'getCommentAuthorName' may produce
'NullPointerException'"

ParaskP7 committed:
Analysis: Guard nullable comment identifier variable on comment detail

Warning: "Argument 'commentIdentifier' might be null"

ParaskP7 committed:
Analysis: Access static method via class color utils reference

Warning: "Static member 'org.wordpress.android.util.ColorUtils.
setImageResourceWithTint(android.widget.ImageView, int, int)' accessed
via instance"

ParaskP7 committed:
Analysis: Suppress same parameter value on comment detail

Warning: "Actual value of parameter 'forceLike' is always 'false'"

FYI: Currently 'likeComment(actionBinding, false)' is used once and
'likeComment(actionBinding, true)' is never used, but this doesn't mean
it will not be used in the future, or wasn't used in the past. As such,
keeping this method signature and suppressing this warning for the time
being seems to be the better approach here.

ParaskP7 committed:
Analysis: Guard nullable comment field on comment detail

Warning: "Argument 'mComment' might be null"

ParaskP7 committed:
Analysis: Remove unnecessary return on comment detail

Warning: "'return' is unnecessary as the last statement in a 'void'
method"

ParaskP7 committed:
Analysis: Simplify if/else for show more menu on comment detail

Warning: "'if' statement can be simplified"

ParaskP7 committed:
Docs: Fixed initialized type on comment detail

ParaskP7 committed:
Analysis: Add missing nl-a to comment field

FYI: 'nl-a' stands for 'nullable annotation'.

PS: This change also removes the below dead code situations:
1. editComment(...) method: The 'mComment == null' that guards the
access to 'intent.putExtra(...)' can never be met. This is due to the
fact that this method was anyway returning on 'mComment' being null.
2. updateStatusViews(...) method: The 'mComment != null' is always true
and thus 'mNote.hasLikedComment()' can never be triggered. Again, this
is due to the fact that this method was anyway returning on 'mComment'
being null.

ParaskP7 committed:
Refactor: Split show comment into two for nullable and non-null comment

ParaskP7 committed:
Docs: Reformat add detail fragment comment

ParaskP7 committed:
Analysis: Use nn-a comment parameter for show/set post title

FYI: 'nn-a' stands for 'non-null annotation'.

ParaskP7 committed:
Analysis: Change to nn-a binding parameter on set post title

FYI: 'nn-a' stands for 'non-null annotation'.

ParaskP7 committed:
Analysis: Use nl-a comment parameter for set reply unique id

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Utilize nl-a on node field

FYI: 'nl-a' stands for 'nullable annotation'.

PS: Contrary to the 'mComment' field, which is usually passed as
'@NonNull' to most methods, the 'mNote' field is usually passes as
'@Nullable' to most methods. This is because a 'note' is only used when
a comment is accessed via the 'Notifications' tab, while a 'comment' is
universal to both, the 'Comments' screen and 'Notifications' tab.

ParaskP7 committed:
Analysis: Add missing nl-a to site field

FYI: 'nl-a' stands for 'nullable annotation'.

PS: This change also remove the 'site' parameter on the
'CommentActions.OnCommentActionListener' interface and its
'onModerateComment(...)' method as it is actually unused. To verify
that look at the 'CommentsDetailActivity' class and its implemetation of
the 'onModerateComment(...)' interface method.

ParaskP7 committed:
Refactor: Reorder site param on top of comment for set comment method

This way any method that is using the 'site', 'comment' and 'note'
parameter will have then in the order specified below:
1. Bindings
2. Site
3. Comment
4. Note
5. Rest

FYI: This change is done for consistency purposes only.

ParaskP7 committed:
Analysis: Add missing nl-a to suggestion adapter field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to suggestion service connection manager fld

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to restored reply text field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to restored note id field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to previous status field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to notifications detail list fragment field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to on post click listener field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to on comment action listener field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to on note comment action listener field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to comment source field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nn-a to enabled actions field

FYI: 'nn-a' stands for 'non-null annotation'.

irfano commented:
> When editing a comment and you end up saving a comment that was already edited with the exact same set of characters in the previous edit, then this new edit is not saved, instead it is being reset to the previous edit.

I tried but couldn't reproduce it.

My smoke tests on comment screens for WP and JP were passed. 🟢 

I couldn't complete reviewing the code. I also saw this comment on [CommentDetailFragment](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java#L130-L131C51): `If you are adding any features or modifying this class, please ping develric or klymyam`. Would they be better reviewers for this PR? I will be AFK this week.

ParaskP7 commented:
👋 @irfano and thanks for starting the review process on this! 🙇 

> I tried but couldn't reproduce it.

Good to hear! 👍 

> My smoke tests on comment screens for WP and JP were passed. 🟢

Awesome, thanks so much for smoke testing this PR! 🙇 

> I couldn't complete reviewing the code. I also saw this comment on [CommentDetailFragment](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java#L130-L131C51): If you are adding any features or modifying this class, please ping develric or klymyam.

Yea, I know, I actually discussed about that with @develric and the rest (see [comment](https://github.com/wordpress-mobile/WordPress-Android/issues/18906#issuecomment-1727322535) and surrounding discussion). From what I understand, although the comment on those classes is saying that, I think for the sake of this work we can still continue with the modification. Also, this code is live in production and, again to my understanding, it will not go away anytime soon. 🤷 

> Would they be better reviewers for this PR?

Maybe they would be better reviewers, but I think every JP/WPAndroid engineer should be able to review this as well. 🤞 

> I will be AFK this week.

I know you are/will be on AFK. I am also on our Apps Infra meetup atm and will be back next week. So, there is no real rush, or urgency in general, that is, in getting this reviewed this week. So please, no worries on that front and thanks for being so thoughtful about it. 🙇

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into refactor/replace-findviewbyid-with-viewbinding-for-comment

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java

ParaskP7 committed:
Cleanup: Remove unused comment and not parameters from edit comment

This was done due to 76381a7e6aa4a90002fffc7458a3aaa01f3ab5e4 and the
fact the unified comments comment edit feature flag is already removed
on 'trunk', causing a merge conflict on this change, which after being
resolved, it left those two unused parameters hanging there.

ParaskP7 commented:
👋 @irfano this is just a friendly ping, are you able to continue with the code review or do you think it is better if I re-assign that to someone else? 🙏

irfano commented:
> 👋 @irfano this is just a friendly ping, are you able to continue with the code review or do you think it is better if I re-assign that to someone else? 🙏

👋🏻 Thank you for the reminder. I was AFK for half of this week as well. However, this PR is on my to-do list, I will review it on Monday.

ParaskP7 commented:
> 👋🏻 Thank you for the reminder. I was AFK for half of this week as well. However, this PR is on my to-do list, I will review it on Monday.

Ah @irfano , I didn't know you were AFK half of this week as well, thinking that you were only AFK last week, apologies for the ping and please take your time with the review, I was just trying to make sure this is on track... 🙏 🙇

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -1242,202 +1369,259 @@ private boolean canShowMore() {
+     * inflated (onCreateView), so we can make sure we set its ID correctly once we have an actual Comment object
+     * to populate it with. Otherwise, we could be searching and finding the container for _another fragment/page
+     * in the viewpager_, which would cause strange results (changing the views for a different fragment than we
+     * intended to).
```
💡I delved into the history of this comment a bit. The comment was [added](https://github.com/wordpress-mobile/WordPress-Android/commit/98c2884338e6498f653a44dc262b68b3aabc8e35) seven years ago. I believe the issue was not being able to use `ChildFragmentManager` because it was available for API 17+. Now, we are using `ChildFragmentManager` in this class, and I think we can safely remove this outdated comment.

ParaskP7 replied:
Thanks so much for this comment @irfano , I agree and this is now done: fcb0288e48311b349de5c851c560803e8d187ac4

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -1447,39 +1631,41 @@ private void onCommentLiked(OnCommentChanged event) {
+                    (coroutineScope, continuation) -> mLocalCommentCacheUpdateHandler.requestCommentsUpdate(
+                            continuation
+                    )
+            );
```
⚠️ `mLocalCommentCacheUpdateHandler.requestCommentsUpdate()` was triggering without any condition before. Now, it is inside `mBinding != null` conditions. There may be a case that binding can be null, we don't trigger `mLocalCommentCacheUpdateHandler.requestCommentsUpdate()`, and that would cause an issue. Perhaps, this won't cause any issues but I noticed a behavior change here. So, to be on the safe side, I would suggest moving this code block out of the bindings' null conditions as before:
```
 BuildersKt.launch(GlobalScope.INSTANCE,
                    Dispatchers.getMain(),
                    CoroutineStart.DEFAULT,
                    (coroutineScope, continuation) -> mLocalCommentCacheUpdateHandler.requestCommentsUpdate(
                            continuation
                    )
            );
```
Same for ` AppLog.i(T.TESTS, "event error type: " + event.error.type + " - message: " + event.error.message);`.

ParaskP7 replied:
Although this might be an edge case, I still think this is a good catch and we better be safe than sorry, thanks for this comment @irfano , done here: bc8ca00904fdad5dba82c2f74a5d714b95554bdf

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -699,178 +720,218 @@ private CommentIdentifier mapCommentIdentifier() {
+    /**
+     * These two views contain all the other views except the progress bar
+     * hide container views when comment is null (will happen when opened from a notification)
+     */
```
💡It would be better to move `These two views contain all the other views except the progress bar
     hide container views when comment is null (will happen when opened from a notification)` comment into the function, just before `binding.nestedScrollView.setVisibility(View.GONE);` I believe the comment is more related with `nestedScrollView` and `layoutBottom`, not related with whole function.

ParaskP7 replied:
Thanks so much for this comment @irfano , I agree and this is now done: 5996160485b31144deb517553efbf7f371a081bd

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -699,178 +720,218 @@ private CommentIdentifier mapCommentIdentifier() {
+            @NonNull CommentActionFooterBinding actionBinding,
+            @NonNull Note note
+    ) {
+        binding.nestedScrollView.setVisibility(View.GONE);
+        binding.layoutBottom.setVisibility(View.GONE);
```
⚠️  These two lines were outside of `if (note != null)` before. If you haven't made this behavior change intentionally, I suggest moving `if (note != null)` condition into `showCommentWhenNullable()`, after these lines.

ParaskP7 replied:
Oh, I think this is a nice catch, it's been a while and I no longer remember why I chose to add the `if (note != null)` at that level (such a complex class). Looking at it again, I think it is indeed better to not introduce any unnecessary behavior change, thanks for this comment @irfano , done here: 962b8d9b053e5bbf81ed8e281a3e12480c9b21b9

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -699,178 +720,218 @@ private CommentIdentifier mapCommentIdentifier() {
 
-        // these two views contain all the other views except the progress bar
-        final View layoutBottom = getView().findViewById(R.id.layout_bottom);
+        if (comment == null) {
+            if (note != null) {
+                showCommentWhenNullable(binding, replyBinding, actionBinding, note);
+            }
+        } else {
+            showCommentWhenNonNull(binding, replyBinding, actionBinding, site, comment, note);
+        }
```
❤️ Good job on shortening the `showComment` function.

irfano reviewed (changes requested):
Thank you for making the reviewer's job easier by creating a well-structured PR. I have not approved the PR to allow you to review my comments.

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into refactor/replace-findviewbyid-with-viewbinding-for-comment

ParaskP7 committed:
Docs: Remove outdated comment content ctr id comment on comment detail

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19232#discussion_r1350567578

ParaskP7 committed:
Refactor: Remove if condition from cache refresh and error log/toast

This is done in order to avoid a potentially behavior change where
requesting a local comment cache refresh would not get triggered if any
of the nullable fields end-up being null at some point.

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19232#discussion_r1350623463

------------------------------------------------------------------------

FYI: In addition to this change and extra the 'site' and 'comment'
condition checks where moved closer to usage. This was done for the very
same reason and to avoid ignoring triggering the like/unlike logic on
cases where the 'site' and 'comment' are null since those field do not
apply there.

ParaskP7 committed:
Docs: Move nested scroll view and layout btm comment on comment detail

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19232#discussion_r1350691492

ParaskP7 committed:
Refactor: Move note check condition after nested scl view and layout btm

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19232#discussion_r1350698850

ParaskP7 commented:
Such a awesome code review @irfano , thank you so much for that and for testing this PR, very much appreciated! 🙇 ❤️ 

FYI: I have just addressed all your comments. Kindly take another look and let me know if there is anything else that needs to be done. 🙏

irfano reviewed (approved):
Thank you for addressing my comments, and I'm adding a minor 🔍, ignorable comment. The patch I suggest includes:
- Renaming `showCommentWhenNullable` to `showCommentWhenNull`,
- Changing the position of the comments to where I believe it fits better.

[Refactor_the_comments_and_the_function_name_in_CommentDetailFragment.patch](https://github.com/wordpress-mobile/WordPress-Android/files/12856037/Refactor_the_comments_and_the_function_name_in_CommentDetailFragment.patch)

I'm approving the PR 🟢. Feel free to apply or ignore may patch, and then you can merge the PR. 💙

ParaskP7 committed:
Docs: Move nested scroll view and layout btm comment on comment detail

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19232#pullrequestreview-1667229816

ParaskP7 committed:
Refactor: Rename show comment when nullable method on comment detail

Related PR Comment: https://github.com/wordpress-mobile/
WordPress-Android/pull/19232#pullrequestreview-1667229816

ParaskP7 commented:
Awesome, thanks so much for approving and sharing the patch @irfano , makes it so much easy to understand exactly what you were referring to! 🙇 ❤️ 

> Renaming showCommentWhenNullable to showCommentWhenNull,

Done here: daefef70b7b217c2262ee9bea58adb7f1e0dad1d

> Changing the position of the comments to where I believe it fits better.

Done here: 1a3bf996dfc4701225af4f7b073e19efc70699fa

🚀
